Data Type Optimization

i-am-sijia opened the pull request:
PR for https://github.com/ActivitySim/activitysim/issues/673

i-am-sijia committed:
trying pandas categorical dtype

i-am-sijia committed:
sweeping out string variables

i-am-sijia committed:
telecommute freq string

i-am-sijia committed:
vehicle type strings

i-am-sijia committed:
categorical is not recognized by numpy

i-am-sijia committed:
df info verbose

i-am-sijia committed:
categoricals for school escorting

i-am-sijia committed:
categoricals for vehicle allocation

i-am-sijia committed:
add condition for pandas categoricals

i-am-sijia committed:
adhoc string variables to categoricals

i-am-sijia committed:
auto downcast for interim variables

i-am-sijia committed:
not use unsigned int

i-am-sijia committed:
numeric overflow on chaperone weights

i-am-sijia committed:
numeric overflow in veh type preprocessor

i-am-sijia committed:
downcast tour variables in source code

i-am-sijia committed:
condition the temps

i-am-sijia committed:
sort time period cat in tour scheduling

i-am-sijia committed:
add loggings

i-am-sijia committed:
formatting

i-am-sijia committed:
update dependencies

i-am-sijia committed:
update unit tests under util\test

i-am-sijia committed:
update unit tests under abm.test

i-am-sijia committed:
Merge branch 'develop' into data-type-op-pd-cat

i-am-sijia committed:
make int downcast an option

i-am-sijia committed:
test moving str comp out of utility evaluation

i-am-sijia committed:
fix mtc and sandag ci tests

i-am-sijia committed:
get categories from alternatives

i-am-sijia committed:
black formatting

i-am-sijia committed:
add `downcast_int` and `downcast_float` into settings

i-am-sijia committed:
fix primary purpose in school escorting

i-am-sijia committed:
sort mode categories

i-am-sijia committed:
fix estimation tests

jpn-- commented on activitysim/core/util.py:
```diff
@@ -323,6 +326,8 @@ def assign_in_place(df, df2):
         assignment left-hand-side (dest)
     df2: pd.DataFrame
         assignment right-hand-side (source)
+    downcast_float: bool
```
include downcast_int also.

jpn-- commented on activitysim/core/util.py:
```diff
@@ -371,6 +379,79 @@ def assign_in_place(df, df2):
+            # if df[col].min() >= 0:
+            #     df[col] = pd.to_numeric(df[col], downcast='unsigned')
+            # else:
+            #     df[col] = pd.to_numeric(df[col], downcast='integer')
```
Looks like we're circumventing this by just downcasting to signed integer regardless.  We don't need this commented _code_ and it can come out, but do please leave/update the contextual comment above it about this topic, so some poor soul doesn't come back in the future and try to save a few bits of RAM by doing this.

i-am-sijia replied:
Yep, I deleted them. I should probably open an issue about the negative values (although it's not really an issue), since it impacts on the data type to use. It would be good to remember this if/when we do the data model.

jpn-- commented on activitysim/abm/models/vehicle_type_choice.py:
```diff
@@ -231,7 +231,9 @@ def construct_model_alternatives(
-        alts_wide.drop(columns="_merge", inplace=True)
+        alts_wide.drop(
+            columns="_merge", inplace=True
+        )  # if this happens, alt_wide length will not match alt_long
```
Considering the comment, is this then a bug that needs to be addressed?

i-am-sijia replied:
Opened issue: https://github.com/ActivitySim/activitysim/issues/789

i-am-sijia replied:
I moved this comment up two lines.

jpn-- commented on activitysim/abm/models/util/test/test_mandatory_tour_frequency.py:
```diff
@@ -56,10 +56,11 @@ def test_mtf():
 
     pdt.assert_series_equal(
-        mandatory_tours.tour_type,
+        mandatory_tours.tour_type.astype(str),
```
instead of dumping to `str`, perhaps we should have the test verify the categorical data type here?

i-am-sijia replied:
I see, they are not pd.cat in this unit test any more. This test calls the method in the util/tour_frequency.py. Initially I implemented the pd.cat conversion in the tour_frequency.py, with hardcoded categories. More recently, I moved the pd.cat conversion to the actual tour frequency models themselves, like mandatory_tour_frequency.py, so that the categories can be automatically set based on the UECs instead of hardcode in .py.

jpn-- commented on activitysim/core/util.py:
```diff
@@ -371,6 +379,79 @@ def assign_in_place(df, df2):
+        if pd.api.types.is_object_dtype(df[c]):
+            df[c] = df[c].astype("category")
+
+    auto_opt_pd_dtypes(df, downcast_int, downcast_float, inplace=True)
```
By running this at the end of the assign-in-place func, it seems like we're first creating a large, inefficient dataframe with poorly optimized data types, and then trying to compress it down by using more efficient types.  In many cases this is probably fine, the data types are not really that big and it's keeping the size down on long-term stored data, but in corner cases this may make the memory usage worse instead of better

i-am-sijia replied:
Correct, auto_opt_pd_dtypes() compresses variables after they are created in, e.g., inputs, annotations, UECs. It's a mask and shortcut rather than changing how activitysim creates variables, i.e., eval().  For this instance, are you suggesting applying auto_opt_pd_dtypes() to df2, rather than df?  When you say "in corner cases", do you mean how pandas converts data types internally and it's possible for it to consume more memory?

jpn-- replied:
When I wrote this I was talking here about how the basic `eval` code creates object type variables, and wouldn't it be better if it didn't.  But I think you're right not to touch that yet, it's likely to get messy and be not-so backwards compatible, which we don't want to do (yet)

jpn-- commented on activitysim/examples/placeholder_psrc/configs/stop_frequency_annotate_tours_preprocessor.csv:
```diff
@@ -45,3 +45,6 @@ AccesibilityADestination off-peak transit,_dest_trOpRetail,"reindex(accessibilit
 AccesibilityAtDestination if transit,pracc,"pracc.where(~tour_mode_is_transit, _dest_trPkRetail.where(_tour_starts_in_peak, _dest_trOpRetail))"
 AccesibilityAtDestination if non_motorized,pracc,"pracc.where(~tour_mode_is_non_motorized, reindex(accessibility.nmRetail, df.destination))"
 ,destination_area_type,"reindex(land_use.area_type, df.destination)"
+,is_mixed,df.composition=='mixed'
```
general comment: the changes we are making in all these model spec and preprocessor files, are they *required* or just helpful?  We don't want to have required changes that injure backwards compatibility.

i-am-sijia replied:
These are required in sharrow mode. It is related to [issue 766](https://github.com/ActivitySim/activitysim/issues/766). I noticed that if a pandas categorical variable is compared to a string, sharrow seems to evaluate it wrong.

jpn-- replied:
ok, I have added 766 to the Phase 9 priority list.

jpn-- reviewed (changes requested):
This is a partial review, there's a lot to go through here and I'm not done but want to share my comments so far instead of holding them all until I'm done.

i-am-sijia committed:
fix typo

i-am-sijia committed:
remove code in comments

i-am-sijia committed:
add check column type in the tests in util/test

i-am-sijia committed:
remove unwanted tracing

@jpn-- I'm gonna remove these files. I used them to write out the memory of each column in the interaction_df. It was useful when I did the diagnosis, but probably not needed in actual implementation.

i-am-sijia committed:
blacken

i-am-sijia committed:
move comment

jpn-- committed:
relax progressive checks for categorical dtypes

jpn-- committed:
update pointers

jpn-- committed:
use empty string as NaN

jpn-- committed:
Merge commit '25e4f6c9055cccaeee2145b2924bdc7d00459abc' into data-type-dev-merge

# Conflicts:
#	activitysim/core/util.py

i-am-sijia committed:
Merge pull request #8 from camsys/data-type-op-pd-cat

relax progressive checks for categorical dtypes

jpn-- committed:
update mtc_ext progressive test for categorical outputs

jpn-- committed:
Merge commit 'ae126f1f85b7ac858f921a762e49e9da534a9efb' into data-type-dev-merge

jpn-- committed:
Merge commit '69465d9534e2ca889ad5a5acaed1eadc9a330a6f' into data-type-dev-merge

# Conflicts:
#	activitysim/core/interaction_simulate.py

i-am-sijia committed:
Merge pull request #9 from camsys/data-type-dev-merge

Merge Develop Branch

i-am-sijia commented:
@jpn-- seems the change of the black version has caused the linting test to fail. I can update my local black and reformat if you are ok with that.

jpn-- commented:
> @jpn-- seems the change of the black version has caused the linting test to fail. I can update my local black and reformat if you are ok with that.

No, I made a mistake and clicked the wrong button to merge code into develop I wasn't intending to.   I rolled it back, should be better if you just merge in the latest changes from the develop branch.

i-am-sijia committed:
Merge pull request #10 from wsp-sag/develop

Bring in the latest Develop

jpn-- committed:
Merge branch 'develop' into data-type-op-pd-cat

# Conflicts:
#	activitysim/abm/models/vehicle_type_choice.py

i-am-sijia committed:
Merge pull request #11 from camsys/data-type-op-pd-cat

latest updates from other PRs
